*Carlos Veny Carmona - 43194073G -* Inteligencia Artificial, UIB

# Práctica FIFA

En esta práctica hay que aplicar las diferentes técnicas vistas en clase (limpieza de datos, analizar el problema y estudiar y ajustar los parámetros del algoritmo de aprendizaje automático) para poder predecir el valor de un jugador sabiendo únicamente las características de un conjunto de jugadores.

## Enlace repositorio

Para realizar el trabajo únicamente se necesitan los fichero `fifa.csv` y `PracticaFIFA` que se encuentra en el siguiente repositorio: https://github.com/carlosveny/practica_IA_fifa.git

## Importar librerias

El primer paso es importar las librerias que usaremos.

In [1]:
import os

from sklearn.model_selection import train_test_split
from sklearn import linear_model
from sklearn.metrics import r2_score

import pandas as pd

## Leer los datos

Usaremos ``pandas`` y una utilidad de la libreria ``os``

In [2]:
df = pd.read_csv(os.path.join("..", "in", "fifa.csv"))
df.head()

Unnamed: 0      ID               Name  Age  \
0           0  158023           L. Messi   31   
1           1   20801  Cristiano Ronaldo   33   
2           2  190871          Neymar Jr   26   
3           3  193080             De Gea   27   
4           4  192985       K. De Bruyne   27   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1   https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2  https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3  https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4  https://cdn.sofifa.org/players/4/19/192985.png     Belgium   

                                  Flag  Overall  Potential  \
0  https://cdn.sofifa.org/flags/52.png       94         94   
1  https://cdn.sofifa.org/flags/38.png       94         94   
2  https://cdn.sofifa.org/flags/54.png       92         93   
3  https://cdn.sofifa.org/flags/45.png       91         93   
4   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club  ... Composure Marking StandingTackle  SlidingTackle  \
0         FC Barcelona  ...      96.0    33.0           28.0           26.0   
1             Juventus  ...      95.0    28.0           31.0           23.0   
2  Paris Saint-Germain  ...      94.0    27.0           24.0           33.0   
3    Manchester United  ...      68.0    15.0           21.0           13.0   
4      Manchester City  ...      88.0    68.0           58.0           51.0   

  GKDiving  GKHandling  GKKicking  GKPositioning GKReflexes Release Clause  
0      6.0        11.0       15.0           14.0        8.0        €226.5M  
1      7.0        11.0       15.0           14.0       11.0        €127.1M  
2      9.0         9.0       15.0           15.0       11.0        €228.1M  
3     90.0        85.0       87.0           88.0       94.0        €138.6M  
4     15.0        13.0        5.0           10.0       13.0        €196.4M  

[5 rows x 89 columns]

## Análisis Datos

Para poder analizar los datos primero tenemos que visualizar todas las columnas del conjunto de datos.

In [3]:
pd.set_option('display.max_columns', None)

In [4]:
df.head()

Unnamed: 0      ID               Name  Age  \
0           0  158023           L. Messi   31   
1           1   20801  Cristiano Ronaldo   33   
2           2  190871          Neymar Jr   26   
3           3  193080             De Gea   27   
4           4  192985       K. De Bruyne   27   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1   https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2  https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3  https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4  https://cdn.sofifa.org/players/4/19/192985.png     Belgium   

                                  Flag  Overall  Potential  \
0  https://cdn.sofifa.org/flags/52.png       94         94   
1  https://cdn.sofifa.org/flags/38.png       94         94   
2  https://cdn.sofifa.org/flags/54.png       92         93   
3  https://cdn.sofifa.org/flags/45.png       91         93   
4   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club                                     Club Logo    Value  \
0         FC Barcelona  https://cdn.sofifa.org/teams/2/light/241.png  €110.5M   
1             Juventus   https://cdn.sofifa.org/teams/2/light/45.png     €77M   
2  Paris Saint-Germain   https://cdn.sofifa.org/teams/2/light/73.png  €118.5M   
3    Manchester United   https://cdn.sofifa.org/teams/2/light/11.png     €72M   
4      Manchester City   https://cdn.sofifa.org/teams/2/light/10.png    €102M   

    Wage  Special Preferred Foot  International Reputation  Weak Foot  \
0  €565K     2202           Left                       5.0        4.0   
1  €405K     2228          Right                       5.0        4.0   
2  €290K     2143          Right                       5.0        5.0   
3  €260K     1471          Right                       4.0        3.0   
4  €355K     2281          Right                       4.0        5.0   

   Skill Moves       Work Rate   Body Type Real Face Position  Jersey Number  \
0          4.0  Medium/ Medium       Messi       Yes       RF           10.0   
1          5.0       High/ Low  C. Ronaldo       Yes       ST            7.0   
2          5.0    High/ Medium      Neymar       Yes       LW           10.0   
3          1.0  Medium/ Medium        Lean       Yes       GK            1.0   
4          4.0      High/ High      Normal       Yes      RCM            7.0   

         Joined Loaned From Contract Valid Until Height  Weight    LS    ST  \
0   Jul 1, 2004         NaN                 2021    5'7  159lbs  88+2  88+2   
1  Jul 10, 2018         NaN                 2022    6'2  183lbs  91+3  91+3   
2   Aug 3, 2017         NaN                 2022    5'9  150lbs  84+3  84+3   
3   Jul 1, 2011         NaN                 2020    6'4  168lbs   NaN   NaN   
4  Aug 30, 2015         NaN                 2023   5'11  154lbs  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     8

A continuación usaremos el comando `describe` para ver estadísticas sobre los datos a analizar.

In [5]:
df.describe()

Unnamed: 0             ID           Age       Overall     Potential  \
count  18207.000000   18207.000000  18207.000000  18207.000000  18207.000000   
mean    9103.000000  214298.338606     25.122206     66.238699     71.307299   
std     5256.052511   29965.244204      4.669943      6.908930      6.136496   
min        0.000000      16.000000     16.000000     46.000000     48.000000   
25%     4551.500000  200315.500000     21.000000     62.000000     67.000000   
50%     9103.000000  221759.000000     25.000000     66.000000     71.000000   
75%    13654.500000  236529.500000     28.000000     71.000000     75.000000   
max    18206.000000  246620.000000     45.000000     94.000000     95.000000   

            Special  International Reputation     Weak Foot   Skill Moves  \
count  18207.000000              18159.000000  18159.000000  18159.000000   
mean    1597.809908                  1.113222      2.947299      2.361308   
std      272.586016                  0.394031      0.660456      0.756164   
min      731.000000                  1.000000      1.000000      1.000000   
25%     1457.000000                  1.000000      3.000000      2.000000   
50%     1635.000000                  1.000000      3.000000      2.000000   
75%     1787.000000                  1.000000      3.000000      3.000000   
max     2346.000000                  5.000000      5.000000      5.000000   

       Jersey Number      Crossing     Finishing  HeadingAccuracy  \
count   18147.000000  18159.000000  18159.000000     18159.000000   
mean       19.546096     49.734181     45.550911        52.298144   
std        15.947765     18.364524     19.525820        17.379909   
min         1.000000      5.000000      2.000000         4.000000   
25%         8.000000     38.000000     30.000000        44.000000   
50%        17.000000     54.000000     49.000000        56.000000   
75%        26.000000     64.000000     62.000000        64.000000   
max        99.000000     93.000000     95.000000        94.000000   

       ShortPassing       Volleys     Dribbling         Curve    FKAccuracy  \
count  18159.000000  18159.000000  18159.000000  18159.000000  18159.000000   
mean      58.686712     42.909026     55.371001     47.170824     42.863153   
std       14.699495     17.694408     18.910371     18.395264     17.478763   
min        7.000000      4.000000      4.000000      6.000000      3.000000   
25%       54.000000     30.000000     49.000000     34.000000     31.000000   
50%       62.000000     44.000000     61.000000     48.000000     41.000000   
75%       68.000000     57.000000     68.000000     62.000000     57.000000   
max       93.000000     90.000000     97.000000     94.000000     94.000000   

        LongPassing   BallControl  Acceleration   SprintSpeed       Agility  \
count  18159.000000  18159.000000  18159.000000  18159.000000  18159.000000   
mean      52.711933     58.369459     64.614076     64.726967     63.503607   
std       15.327870     16.686595     14.927780     14.649953     14.766049   
min        9.000000      5.000000     12.000000     12.000000     14.000000   
25%       43.000000     54.000000     57.000000     57.000000     55.000000   
50%       56.000000     63.000000     67.000000     67.000000     66.000000   
75%       64.000000     69.000000     75.000000     75.000000     74.000000   
max       93.000000     96.000000     97.000000     96.000000     96.000000   

          Reactions       Balance     ShotPower       Jumping       Stamina  \
count  18159.000000  18159.000000  18159.000000  18159.000000  18159.000000   
mean      61.836610     63.966573     55.460047     65.089432     63.219946   
std        9.010464     14.136166     17.237958     11.820044     15.894741   
min       21.000000     16.000000      2.000000     15.000000     12.000000   
25%       56.000000     56.000000     45.000000     58.000000     56.000000   
50%       62.000000     66.000000     59.000000     66.000000     66.00000

## Limpieza de datos

Gracias al comando usado anteriormente podemos ver que las dos primeras columnas no nos aportan información importante ya que son un contador y el identificador del jugador asi que procedemos a eliminarlas.

In [6]:
df = df.iloc[:, 2:]
df.head()

Name  Age                                           Photo  \
0           L. Messi   31  https://cdn.sofifa.org/players/4/19/158023.png   
1  Cristiano Ronaldo   33   https://cdn.sofifa.org/players/4/19/20801.png   
2          Neymar Jr   26  https://cdn.sofifa.org/players/4/19/190871.png   
3             De Gea   27  https://cdn.sofifa.org/players/4/19/193080.png   
4       K. De Bruyne   27  https://cdn.sofifa.org/players/4/19/192985.png   

  Nationality                                 Flag  Overall  Potential  \
0   Argentina  https://cdn.sofifa.org/flags/52.png       94         94   
1    Portugal  https://cdn.sofifa.org/flags/38.png       94         94   
2      Brazil  https://cdn.sofifa.org/flags/54.png       92         93   
3       Spain  https://cdn.sofifa.org/flags/45.png       91         93   
4     Belgium   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club                                     Club Logo    Value  \
0         FC Barcelona  https://cdn.sofifa.org/teams/2/light/241.png  €110.5M   
1             Juventus   https://cdn.sofifa.org/teams/2/light/45.png     €77M   
2  Paris Saint-Germain   https://cdn.sofifa.org/teams/2/light/73.png  €118.5M   
3    Manchester United   https://cdn.sofifa.org/teams/2/light/11.png     €72M   
4      Manchester City   https://cdn.sofifa.org/teams/2/light/10.png    €102M   

    Wage  Special Preferred Foot  International Reputation  Weak Foot  \
0  €565K     2202           Left                       5.0        4.0   
1  €405K     2228          Right                       5.0        4.0   
2  €290K     2143          Right                       5.0        5.0   
3  €260K     1471          Right                       4.0        3.0   
4  €355K     2281          Right                       4.0        5.0   

   Skill Moves       Work Rate   Body Type Real Face Position  Jersey Number  \
0          4.0  Medium/ Medium       Messi       Yes       RF           10.0   
1          5.0       High/ Low  C. Ronaldo       Yes       ST            7.0   
2          5.0    High/ Medium      Neymar       Yes       LW           10.0   
3          1.0  Medium/ Medium        Lean       Yes       GK            1.0   
4          4.0      High/ High      Normal       Yes      RCM            7.0   

         Joined Loaned From Contract Valid Until Height  Weight    LS    ST  \
0   Jul 1, 2004         NaN                 2021    5'7  159lbs  88+2  88+2   
1  Jul 10, 2018         NaN                 2022    6'2  183lbs  91+3  91+3   
2   Aug 3, 2017         NaN                 2022    5'9  150lbs  84+3  84+3   
3   Jul 1, 2011         NaN                 2020    6'4  168lbs   NaN   NaN   
4  Aug 30, 2015         NaN                 2023   5'11  154lbs  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     86.0       97.0   
1      84.0       94.0             89.0          81.0     87.0       88.0   
2      79.0       87.0             62.0          84.0     84.0       

Algunos valores como la foto del jugador, la bandera, el logo del club... no aportan información relevante que nos pueda ayudar a predecir el valor de un jugador. Por ello realizamos una eliminación de las siguientes columnas del conjunto de datos:
- Name
- Photo
- Flag
- Club Logo
- Special
- Work Rate
- Real Face
- Jersey Number
- Joined
- Loaned From
- Contract Valid Until

In [7]:
pd.options.mode.chained_assignment = None
df = df.drop(columns=['Name', 'Photo', 'Flag', 'Club Logo', 'Special', 'Work Rate', 'Real Face', 'Jersey Number', 'Joined', 'Loaned From', 'Contract Valid Until'])
df.head()

Age Nationality  Overall  Potential                 Club    Value   Wage  \
0   31   Argentina       94         94         FC Barcelona  €110.5M  €565K   
1   33    Portugal       94         94             Juventus     €77M  €405K   
2   26      Brazil       92         93  Paris Saint-Germain  €118.5M  €290K   
3   27       Spain       91         93    Manchester United     €72M  €260K   
4   27     Belgium       91         92      Manchester City    €102M  €355K   

  Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
0           Left                       5.0        4.0          4.0   
1          Right                       5.0        4.0          5.0   
2          Right                       5.0        5.0          5.0   
3          Right                       4.0        3.0          1.0   
4          Right                       4.0        5.0          4.0   

    Body Type Position Height  Weight    LS    ST    RS    LW    LF    CF  \
0       Messi       RF    5'7  159lbs  88+2  88+2  88+2  92+2  93+2  93+2   
1  C. Ronaldo       ST    6'2  183lbs  91+3  91+3  91+3  89+3  90+3  90+3   
2      Neymar       LW    5'9  150lbs  84+3  84+3  84+3  89+3  89+3  89+3   
3        Lean       GK    6'4  168lbs   NaN   NaN   NaN   NaN   NaN   NaN   
4      Normal      RCM   5'11  154lbs  82+3  82+3  82+3  87+3  87+3  87+3   

     RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM   LWB   LDM  \
0  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2   
1  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3   
2  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3   

    CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  Finishing  \
0  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0       95.0   
1  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0       94.0   
2  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3      79.0       87.0   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN      17.0       13.0   
4  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3      93.0       82.0   

   HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  \
0             70.0          90.0     86.0       97.0   93.0        94.0   
1             89.0          81.0     87.0       88.0   81.0        76.0   
2             62.0          84.0     84.0       96.0   88.0        87.0   
3             21.0          50.0     13.0       18.0   21.0        19.0   
4             55.0          92.0     82.0       86.0   85.0        83.0   

   LongPassing  BallControl  Acceleration  SprintSpeed  Agility  Reactions  \
0         87.0         96.0          91.0         86.0     91.0       95.0   
1         77.0         94.0          89.0         91.0     87.0       96.0   
2         78.0         95.0          94.0         90.0     96.0       94.0   
3         51.0         42.0          57.0         58.0     60.0       90.0   
4         91.0         91.0          78.0         76.0     79.0       91.0   

   Balance  ShotPower  Jumping  Stamina  Strength  LongShots  Aggression  \
0     95.0       85.0     68.0     72.0      59.0       94.0        48.0   
1     70.0       95.0     95.0     88.0      79.0       93.0        63.0   
2     84.0       80.0     61.0     81.0      49.0       82.0        56.0   
3     43.0       31.0     67.0     43.0      64.0       12.0        38.0   
4     77.0       91.0     63.0     90.0      75.0       91.0        76.0   

   Interceptions  Positioning  Vision  Penalties  Composure  Marking  \
0           22.0         94.0    94.0       75.0       96.0     33.0   
1           29.0         95.0    82.0       85.0       95.0     28.0   
2           36.0         89.0    87.0       81.0       94.0     27.0   
3           30.0         12.0    68.0       40.0       68.0     15.0   

## Gestión de *NaN*

Una vez eliminadas las columnas que no nos hacen falta podemos ver que los datos contienen *NaN* (Not a Number)
Para saber donde se localizan los *NaN* podemos usar el comando `isna`.

In [8]:
df[df.isna().any(axis=1)].head() #mostrar columnas con NaN

Age Nationality  Overall  Potential               Club   Value   Wage  \
3    27       Spain       91         93  Manchester United    €72M  €260K   
9    25    Slovenia       90         93    Atlético Madrid    €68M   €94K   
18   26     Germany       89         92       FC Barcelona    €58M  €240K   
19   26     Belgium       89         90        Real Madrid  €53.5M  €240K   
22   32     Germany       89         89  FC Bayern München    €38M  €130K   

   Preferred Foot  International Reputation  Weak Foot  Skill Moves Body Type  \
3           Right                       4.0        3.0          1.0      Lean   
9           Right                       3.0        3.0          1.0    Normal   
18          Right                       3.0        4.0          1.0    Normal   
19           Left                       4.0        2.0          1.0  Courtois   
22          Right                       5.0        4.0          1.0    Normal   

   Position Height  Weight   LS   ST   RS   LW   LF   CF   RF   RW  LAM  CAM  \
3        GK    6'4  168lbs  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
9        GK    6'2  192lbs  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18       GK    6'2  187lbs  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
19       GK    6'6  212lbs  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
22       GK    6'4  203lbs  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

    RAM   LM  LCM   CM  RCM   RM  LWB  LDM  CDM  RDM  RWB   LB  LCB   CB  RCB  \
3   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
9   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
19  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
22  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

     RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
3   NaN      17.0       13.0             21.0          50.0     13.0   
9   NaN      13.0       11.0             15.0          29.0     13.0   
18  NaN      15.0       14.0             11.0          36.0     14.0   
19  NaN      14.0       14.0             13.0          33.0     12.0   
22  NaN      15.0       13.0             25.0          55.0     11.0   

    Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
3        18.0   21.0        19.0         51.0         42.0          57.0   
9        12.0   13.0        14.0         26.0         16.0          43.0   
18       17.0   18.0        12.0         42.0         18.0          38.0   
19       13.0   19.0        20.0         35.0         23.0          46.0   
22       30.0   14.0        11.0         59.0         48.0          54.0   

    SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
3          58.0     60.0       90.0     43.0       31.0     67.0     43.0   
9          60.0     67.0       86.0     49.0       22.0     76.0     41.0   
18         50.0     37.0       85.0     43.0       22.0     79.0     35.0   
19         52.0     61.0       84.0     45.0       36.0     68.0     38.0   
22         60.0     51.0       84.0     35.0       25.0     77.0     43.0   

    Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
3       64.0       12.0        38.0           30.0         12.0    68.0   
9       78.0       12.0        34.0           19.0         11.0    70.0   
18      79.0       10.0        43.0           22.0         11.0    69.0   
19      70.0       17.0        23.0           15.0         13.0    44.0   
22      80.0       16.0        29.0           30.0         12.0    70.0   

    Penalties  Composure  Marking  StandingTackle  SlidingTackle  GKDiving  \
3        40.0       68.0     15.0            21.0           13.0      90.0   
9        11.0       70.0     27.0            12.0           18.0      86.0   
18       25.0       69.0     25.0            13.

Gracias a los comandos anteriores vemos que hay 76 columnas con *NaN*. Para saber exactamente cuantos *NaN* hay en cada columna podemos usar el comando `sum` de la siguiente manera:

In [9]:
pd.set_option('display.max_rows', None)
df.isna().sum()

Age                            0
Nationality                    0
Overall                        0
Potential                      0
Club                         241
Value                          0
Wage                           0
Preferred Foot                48
International Reputation      48
Weak Foot                     48
Skill Moves                   48
Body Type                     48
Position                      60
Height                        48
Weight                        48
LS                          2085
ST                          2085
RS                          2085
LW                          2085
LF                          2085
CF                          2085
RF                          2085
RW                          2085
LAM                         2085
CAM                         2085
RAM                         2085
LM                          2085
LCM                         2085
CM                          2085
RCM                         2085
RM        

Gracias a la lista anterior podemos ver que hay muchas columnas que tienen el mismo número de *NaN* que otras. Esto significa que muy probablemente esos *NaN* pertenezcan a los mismos jugadores.

Por ello vamos a dividir la gestión de los *NaN* segun las siguientes columnas principales y despues comprobaremos si nos ha quedado algun *NaN* por gestionar:
- Club
- Preferred Foot
- Release Clause
- LS

### Club
Para ver que jugadores no tienen **Club** usamos el siguiente comando:

In [10]:
pd.set_option('display.max_rows', 10)
df.loc[df.Club != df.Club].head()

Age Nationality  Overall  Potential Club Value Wage Preferred Foot  \
452   24   Argentina       80         85  NaN    €0   €0          Right   
538   33      Sweden       80         80  NaN    €0   €0          Right   
568   26      Russia       79         81  NaN    €0   €0          Right   
677   29      Russia       79         79  NaN    €0   €0          Right   
874   29      Russia       78         78  NaN    €0   €0          Right   

     International Reputation  Weak Foot  Skill Moves Body Type Position  \
452                       2.0        4.0          4.0    Normal       CM   
538                       2.0        4.0          2.0    Normal      LCB   
568                       1.0        3.0          1.0    Normal       GK   
677                       2.0        3.0          3.0      Lean       RB   
874                       2.0        3.0          3.0    Stocky       ST   

    Height  Weight    LS    ST    RS    LW    LF    CF    RF    RW   LAM  \
452   5'11  165lbs  71+2  71+2  71+2  75+2  75+2  75+2  75+2  75+2  77+2   
538    6'4  185lbs  62+2  62+2  62+2  56+2  58+2  58+2  58+2  56+2  58+2   
568    6'2  176lbs   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
677   5'10  154lbs  70+2  70+2  70+2  73+2  72+2  72+2  72+2  73+2  73+2   
874    6'5  201lbs  77+2  77+2  77+2  71+2  74+2  74+2  74+2  71+2  71+2   

      CAM   RAM    LM   LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB  \
452  77+2  77+2  76+2  79+2  79+2  79+2  76+2  75+2  77+2  77+2  77+2  75+2   
538  58+2  58+2  57+2  64+2  64+2  64+2  57+2  68+2  74+2  74+2  74+2  68+2   
568   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
677  73+2  73+2  75+2  74+2  74+2  74+2  75+2  78+2  75+2  75+2  75+2  78+2   
874  71+2  71+2  71+2  66+2  66+2  66+2  71+2  52+2  52+2  52+2  52+2  52+2   

       LB   LCB    CB   RCB    RB  Crossing  Finishing  HeadingAccuracy  \
452  74+2  72+2  72+2  72+2  74+2      76.0       55.0             60.0   
538  70+2  79+2  79+2  79+2  70+2      49.0       51.0             81.0   
568   NaN   NaN   NaN   NaN   NaN      16.0       14.0             17.0   
677  78+2  73+2  73+2  73+2  78+2      73.0       61.0             69.0   
874  48+2  48+2  48+2  48+2  48+2      61.0       79.0             86.0   

     ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  LongPassing  \
452          84.0     73.0       78.0   79.0        78.0         82.0   
538          73.0     37.0       49.0   36.0        40.0         67.0   
568          25.0     13.0       15.0   18.0        17.0         32.0   
677          79.0     57.0       72.0   49.0        46.0         75.0   
874          71.0     74.0       71.0   64.0        60.0         55.0   

     BallControl  Acceleration  SprintSpeed  Agility  Reactions  Balance  \
452         82.0          75.0         69.0     77.0       74.0     77.0   
538         63.0          46.0         49.0     55.0       76.0     36.0   
568         17.0          58.0         54.0     36.0       76.0     50.0   
677         72.0          84.0         90.0     80.0       75.0     76.0   
874         77.0          66.0         65.0     50.0       75.0     32.0   

     ShotPower  Jumping  Stamina  Strength  LongShots  Aggression  \
452       82.0     61.0     79.0      69.0       80.0        79.0   
538       74.0     64.0     67.0      83.0       59.0        81.0   
568       24.0     60.0     27.0      70.0       13.0        26.0   
677       67.0     85.0     93.0      68.0       57.0        65.0   
874       78.0     63.0     77.0      93.0       68.0        75.0   

     Interceptions  Positioning  Vision  Penalties  Composure  Marking  \
452           72.0         74.0    82.0       57.0       74.0     73.0   
538           82.0         54.0    49.0       79.0       78.0     82.0   
568           20.0         11.0    63.0       15.0       69.0     18.0   
677           71.0         77.0    72.0       41.0       73.0     76.0   
874           30.0         78.0  

Estos jugadores no tienen Club porque su Club no se encuentra dentro del FIFA, ya que hay determinados Clubs pequeños que no están en el juego. Vemos que solo son 241 jugadores por lo que podemos eliminarlos del conjunto de datos.

In [11]:
df = df.dropna(subset = ['Club'])
df.head()

Age Nationality  Overall  Potential                 Club    Value   Wage  \
0   31   Argentina       94         94         FC Barcelona  €110.5M  €565K   
1   33    Portugal       94         94             Juventus     €77M  €405K   
2   26      Brazil       92         93  Paris Saint-Germain  €118.5M  €290K   
3   27       Spain       91         93    Manchester United     €72M  €260K   
4   27     Belgium       91         92      Manchester City    €102M  €355K   

  Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
0           Left                       5.0        4.0          4.0   
1          Right                       5.0        4.0          5.0   
2          Right                       5.0        5.0          5.0   
3          Right                       4.0        3.0          1.0   
4          Right                       4.0        5.0          4.0   

    Body Type Position Height  Weight    LS    ST    RS    LW    LF    CF  \
0       Messi       RF    5'7  159lbs  88+2  88+2  88+2  92+2  93+2  93+2   
1  C. Ronaldo       ST    6'2  183lbs  91+3  91+3  91+3  89+3  90+3  90+3   
2      Neymar       LW    5'9  150lbs  84+3  84+3  84+3  89+3  89+3  89+3   
3        Lean       GK    6'4  168lbs   NaN   NaN   NaN   NaN   NaN   NaN   
4      Normal      RCM   5'11  154lbs  82+3  82+3  82+3  87+3  87+3  87+3   

     RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM   LWB   LDM  \
0  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2   
1  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3   
2  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3   

    CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  Finishing  \
0  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0       95.0   
1  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0       94.0   
2  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3      79.0       87.0   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN      17.0       13.0   
4  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3      93.0       82.0   

   HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  \
0             70.0          90.0     86.0       97.0   93.0        94.0   
1             89.0          81.0     87.0       88.0   81.0        76.0   
2             62.0          84.0     84.0       96.0   88.0        87.0   
3             21.0          50.0     13.0       18.0   21.0        19.0   
4             55.0          92.0     82.0       86.0   85.0        83.0   

   LongPassing  BallControl  Acceleration  SprintSpeed  Agility  Reactions  \
0         87.0         96.0          91.0         86.0     91.0       95.0   
1         77.0         94.0          89.0         91.0     87.0       96.0   
2         78.0         95.0          94.0         90.0     96.0       94.0   
3         51.0         42.0          57.0         58.0     60.0       90.0   
4         91.0         91.0          78.0         76.0     79.0       91.0   

   Balance  ShotPower  Jumping  Stamina  Strength  LongShots  Aggression  \
0     95.0       85.0     68.0     72.0      59.0       94.0        48.0   
1     70.0       95.0     95.0     88.0      79.0       93.0        63.0   
2     84.0       80.0     61.0     81.0      49.0       82.0        56.0   
3     43.0       31.0     67.0     43.0      64.0       12.0        38.0   
4     77.0       91.0     63.0     90.0      75.0       91.0        76.0   

   Interceptions  Positioning  Vision  Penalties  Composure  Marking  \
0           22.0         94.0    94.0       75.0       96.0     33.0   
1           29.0         95.0    82.0       85.0       95.0     28.0   
2           36.0         89.0    87.0       81.0       94.0     27.0   
3           30.0         12.0    68.0       40.0       68.0     15.0   

### Preferred Foot
Siguiendo la lista de *NaN* mostrada anteriormente se puede ver que solo hay 48 jugadores sin este atributo por lo que también podemos eliminarlos de la siguiente manera:

In [12]:
df = df.dropna(subset = ['Preferred Foot'])
df.head()

Age Nationality  Overall  Potential                 Club    Value   Wage  \
0   31   Argentina       94         94         FC Barcelona  €110.5M  €565K   
1   33    Portugal       94         94             Juventus     €77M  €405K   
2   26      Brazil       92         93  Paris Saint-Germain  €118.5M  €290K   
3   27       Spain       91         93    Manchester United     €72M  €260K   
4   27     Belgium       91         92      Manchester City    €102M  €355K   

  Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
0           Left                       5.0        4.0          4.0   
1          Right                       5.0        4.0          5.0   
2          Right                       5.0        5.0          5.0   
3          Right                       4.0        3.0          1.0   
4          Right                       4.0        5.0          4.0   

    Body Type Position Height  Weight    LS    ST    RS    LW    LF    CF  \
0       Messi       RF    5'7  159lbs  88+2  88+2  88+2  92+2  93+2  93+2   
1  C. Ronaldo       ST    6'2  183lbs  91+3  91+3  91+3  89+3  90+3  90+3   
2      Neymar       LW    5'9  150lbs  84+3  84+3  84+3  89+3  89+3  89+3   
3        Lean       GK    6'4  168lbs   NaN   NaN   NaN   NaN   NaN   NaN   
4      Normal      RCM   5'11  154lbs  82+3  82+3  82+3  87+3  87+3  87+3   

     RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM   LWB   LDM  \
0  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2   
1  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3   
2  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3   

    CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  Finishing  \
0  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0       95.0   
1  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0       94.0   
2  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3      79.0       87.0   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN      17.0       13.0   
4  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3      93.0       82.0   

   HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  \
0             70.0          90.0     86.0       97.0   93.0        94.0   
1             89.0          81.0     87.0       88.0   81.0        76.0   
2             62.0          84.0     84.0       96.0   88.0        87.0   
3             21.0          50.0     13.0       18.0   21.0        19.0   
4             55.0          92.0     82.0       86.0   85.0        83.0   

   LongPassing  BallControl  Acceleration  SprintSpeed  Agility  Reactions  \
0         87.0         96.0          91.0         86.0     91.0       95.0   
1         77.0         94.0          89.0         91.0     87.0       96.0   
2         78.0         95.0          94.0         90.0     96.0       94.0   
3         51.0         42.0          57.0         58.0     60.0       90.0   
4         91.0         91.0          78.0         76.0     79.0       91.0   

   Balance  ShotPower  Jumping  Stamina  Strength  LongShots  Aggression  \
0     95.0       85.0     68.0     72.0      59.0       94.0        48.0   
1     70.0       95.0     95.0     88.0      79.0       93.0        63.0   
2     84.0       80.0     61.0     81.0      49.0       82.0        56.0   
3     43.0       31.0     67.0     43.0      64.0       12.0        38.0   
4     77.0       91.0     63.0     90.0      75.0       91.0        76.0   

   Interceptions  Positioning  Vision  Penalties  Composure  Marking  \
0           22.0         94.0    94.0       75.0       96.0     33.0   
1           29.0         95.0    82.0       85.0       95.0     28.0   
2           36.0         89.0    87.0       81.0       94.0     27.0   
3           30.0         12.0    68.0       40.0       68.0     15.0   

### Release Clause
Hay muchos jugadores que no disponen de cláusula de liberación. Para no tener que eliminar a estos jugadores se les asigna en 0€ el valor de su cláusula de liberación de la siguiente manera:

In [13]:
df["Release Clause"] = df["Release Clause"].fillna(0)
df.head()

Age Nationality  Overall  Potential                 Club    Value   Wage  \
0   31   Argentina       94         94         FC Barcelona  €110.5M  €565K   
1   33    Portugal       94         94             Juventus     €77M  €405K   
2   26      Brazil       92         93  Paris Saint-Germain  €118.5M  €290K   
3   27       Spain       91         93    Manchester United     €72M  €260K   
4   27     Belgium       91         92      Manchester City    €102M  €355K   

  Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
0           Left                       5.0        4.0          4.0   
1          Right                       5.0        4.0          5.0   
2          Right                       5.0        5.0          5.0   
3          Right                       4.0        3.0          1.0   
4          Right                       4.0        5.0          4.0   

    Body Type Position Height  Weight    LS    ST    RS    LW    LF    CF  \
0       Messi       RF    5'7  159lbs  88+2  88+2  88+2  92+2  93+2  93+2   
1  C. Ronaldo       ST    6'2  183lbs  91+3  91+3  91+3  89+3  90+3  90+3   
2      Neymar       LW    5'9  150lbs  84+3  84+3  84+3  89+3  89+3  89+3   
3        Lean       GK    6'4  168lbs   NaN   NaN   NaN   NaN   NaN   NaN   
4      Normal      RCM   5'11  154lbs  82+3  82+3  82+3  87+3  87+3  87+3   

     RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM   LWB   LDM  \
0  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2   
1  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3   
2  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3   

    CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  Finishing  \
0  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0       95.0   
1  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0       94.0   
2  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3      79.0       87.0   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN      17.0       13.0   
4  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3      93.0       82.0   

   HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  \
0             70.0          90.0     86.0       97.0   93.0        94.0   
1             89.0          81.0     87.0       88.0   81.0        76.0   
2             62.0          84.0     84.0       96.0   88.0        87.0   
3             21.0          50.0     13.0       18.0   21.0        19.0   
4             55.0          92.0     82.0       86.0   85.0        83.0   

   LongPassing  BallControl  Acceleration  SprintSpeed  Agility  Reactions  \
0         87.0         96.0          91.0         86.0     91.0       95.0   
1         77.0         94.0          89.0         91.0     87.0       96.0   
2         78.0         95.0          94.0         90.0     96.0       94.0   
3         51.0         42.0          57.0         58.0     60.0       90.0   
4         91.0         91.0          78.0         76.0     79.0       91.0   

   Balance  ShotPower  Jumping  Stamina  Strength  LongShots  Aggression  \
0     95.0       85.0     68.0     72.0      59.0       94.0        48.0   
1     70.0       95.0     95.0     88.0      79.0       93.0        63.0   
2     84.0       80.0     61.0     81.0      49.0       82.0        56.0   
3     43.0       31.0     67.0     43.0      64.0       12.0        38.0   
4     77.0       91.0     63.0     90.0      75.0       91.0        76.0   

   Interceptions  Positioning  Vision  Penalties  Composure  Marking  \
0           22.0         94.0    94.0       75.0       96.0     33.0   
1           29.0         95.0    82.0       85.0       95.0     28.0   
2           36.0         89.0    87.0       81.0       94.0     27.0   
3           30.0         12.0    68.0       40.0       68.0     15.0   

### LS
Del mismo modo, hay 2085 jugadores con *NaN* en esta columna (según la lista anterior, ya que puede que ya hayamos eliminado algunos con los comandos anteriores). Para poder entender que tipo de jugadores son debemos ejecutar un comando que nos muestre todos los jugadores con *NaN* en la columna **LS**:

In [14]:
df.loc[df.LS != df.LS].head()

Age Nationality  Overall  Potential               Club   Value   Wage  \
3    27       Spain       91         93  Manchester United    €72M  €260K   
9    25    Slovenia       90         93    Atlético Madrid    €68M   €94K   
18   26     Germany       89         92       FC Barcelona    €58M  €240K   
19   26     Belgium       89         90        Real Madrid  €53.5M  €240K   
22   32     Germany       89         89  FC Bayern München    €38M  €130K   

   Preferred Foot  International Reputation  Weak Foot  Skill Moves Body Type  \
3           Right                       4.0        3.0          1.0      Lean   
9           Right                       3.0        3.0          1.0    Normal   
18          Right                       3.0        4.0          1.0    Normal   
19           Left                       4.0        2.0          1.0  Courtois   
22          Right                       5.0        4.0          1.0    Normal   

   Position Height  Weight   LS   ST   RS   LW   LF   CF   RF   RW  LAM  CAM  \
3        GK    6'4  168lbs  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
9        GK    6'2  192lbs  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18       GK    6'2  187lbs  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
19       GK    6'6  212lbs  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
22       GK    6'4  203lbs  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

    RAM   LM  LCM   CM  RCM   RM  LWB  LDM  CDM  RDM  RWB   LB  LCB   CB  RCB  \
3   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
9   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
19  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
22  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

     RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
3   NaN      17.0       13.0             21.0          50.0     13.0   
9   NaN      13.0       11.0             15.0          29.0     13.0   
18  NaN      15.0       14.0             11.0          36.0     14.0   
19  NaN      14.0       14.0             13.0          33.0     12.0   
22  NaN      15.0       13.0             25.0          55.0     11.0   

    Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
3        18.0   21.0        19.0         51.0         42.0          57.0   
9        12.0   13.0        14.0         26.0         16.0          43.0   
18       17.0   18.0        12.0         42.0         18.0          38.0   
19       13.0   19.0        20.0         35.0         23.0          46.0   
22       30.0   14.0        11.0         59.0         48.0          54.0   

    SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
3          58.0     60.0       90.0     43.0       31.0     67.0     43.0   
9          60.0     67.0       86.0     49.0       22.0     76.0     41.0   
18         50.0     37.0       85.0     43.0       22.0     79.0     35.0   
19         52.0     61.0       84.0     45.0       36.0     68.0     38.0   
22         60.0     51.0       84.0     35.0       25.0     77.0     43.0   

    Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
3       64.0       12.0        38.0           30.0         12.0    68.0   
9       78.0       12.0        34.0           19.0         11.0    70.0   
18      79.0       10.0        43.0           22.0         11.0    69.0   
19      70.0       17.0        23.0           15.0         13.0    44.0   
22      80.0       16.0        29.0           30.0         12.0    70.0   

    Penalties  Composure  Marking  StandingTackle  SlidingTackle  GKDiving  \
3        40.0       68.0     15.0            21.0           13.0      90.0   
9        11.0       70.0     27.0            12.0           18.0      86.0   
18       25.0       69.0     25.0            13.

Con una primera visión deducimos que se trata de porteros y que no tienen muchos de los atributos que sí tienen los jugadores. Para facilitar el tratamiento de los datos procedemos a sustituir los *NaN* por 0 y así no eliminaremos a tantos jugadores del conjunto de datos.

Son muchas columnas sobre las que tenemos que aplicar la función `fillna(0)` y para no ir una por una primero hay que saber si solo nos queda gestionar los *NaN* de las columnas asociadas al atributo **LS**. Para ello listamos las columnas que contienen *NaN*:

In [15]:
pd.set_option('display.max_rows', None)
df.columns[df.isna().any()].tolist()

['LS',
 'ST',
 'RS',
 'LW',
 'LF',
 'CF',
 'RF',
 'RW',
 'LAM',
 'CAM',
 'RAM',
 'LM',
 'LCM',
 'CM',
 'RCM',
 'RM',
 'LWB',
 'LDM',
 'CDM',
 'RDM',
 'RWB',
 'LB',
 'LCB',
 'CB',
 'RCB',
 'RB']

Vemos que solo quedan las columnas asociadas a **LS** por lo que podemos aplicar la función `fillna(0)` sobre todo el conjunto de datos.

In [16]:
pd.set_option('display.max_rows', 10)
df = df.fillna(0)

Para acabar la gestión de los *NaN* podemos volver a ejecutar el comando `isna` que nos indica en que columnas tenemos valores vacios. Nos debería salir en blanco ya que no debería quedar ningun *NaN* en el conjunto de datos.

In [17]:
df.columns[df.isna().any()].tolist()

[]

## Gestión de unidades

Una vez solucionado el problema de los *NaN* para todas las columnas tenemos el problema de las unidades en el caso de las columnas que son valores. Estos valores son una cadena de texto (string) y necesitamos convertirlos a un número para que podamos trabajar con ellos.

Para conseguir esto definimos una función que pasa las cadenas de texto con notación *K* o *M* a un float.

In [18]:
def value_to_float(x):
    """
    From K and M to float.
    
    """
    if type(x) == float or type(x) == int:
        return x
    x = x.replace('€', '')
    ret_val = 0.0
    
    if type(x) == float or type(x) == int:
        ret_val = x
    if 'K' in x:
        if len(x) > 1:
            ret_val = float(x.replace('K', ''))
        ret_val = ret_val *1000
    if 'M' in x:
        if len(x) > 1:
            ret_val = float(x.replace('M', ''))
        ret_val = ret_val * 1000000.0
    return ret_val

Ahora ya podemos aplicar esta función a las columnas pertinentes. En este caso son: **Value**, **Wage** y **Release Clause**.

In [19]:
df["Value"] = df["Value"].apply(value_to_float)
df["Wage"] = df["Wage"].apply(value_to_float)
df["Release Clause"] = df["Release Clause"].apply(value_to_float)
df.head()

Age Nationality  Overall  Potential                 Club        Value  \
0   31   Argentina       94         94         FC Barcelona  110500000.0   
1   33    Portugal       94         94             Juventus   77000000.0   
2   26      Brazil       92         93  Paris Saint-Germain  118500000.0   
3   27       Spain       91         93    Manchester United   72000000.0   
4   27     Belgium       91         92      Manchester City  102000000.0   

       Wage Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
0  565000.0           Left                       5.0        4.0          4.0   
1  405000.0          Right                       5.0        4.0          5.0   
2  290000.0          Right                       5.0        5.0          5.0   
3  260000.0          Right                       4.0        3.0          1.0   
4  355000.0          Right                       4.0        5.0          4.0   

    Body Type Position Height  Weight    LS    ST    RS    LW    LF    CF  \
0       Messi       RF    5'7  159lbs  88+2  88+2  88+2  92+2  93+2  93+2   
1  C. Ronaldo       ST    6'2  183lbs  91+3  91+3  91+3  89+3  90+3  90+3   
2      Neymar       LW    5'9  150lbs  84+3  84+3  84+3  89+3  89+3  89+3   
3        Lean       GK    6'4  168lbs     0     0     0     0     0     0   
4      Normal      RCM   5'11  154lbs  82+3  82+3  82+3  87+3  87+3  87+3   

     RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM   LWB   LDM  \
0  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2   
1  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3   
2  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3   
3     0     0     0     0     0     0     0     0     0     0     0     0   
4  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3   

    CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  Finishing  \
0  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0       95.0   
1  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0       94.0   
2  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3      79.0       87.0   
3     0     0     0     0     0     0     0     0      17.0       13.0   
4  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3      93.0       82.0   

   HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  \
0             70.0          90.0     86.0       97.0   93.0        94.0   
1             89.0          81.0     87.0       88.0   81.0        76.0   
2             62.0          84.0     84.0       96.0   88.0        87.0   
3             21.0          50.0     13.0       18.0   21.0        19.0   
4             55.0          92.0     82.0       86.0   85.0        83.0   

   LongPassing  BallControl  Acceleration  SprintSpeed  Agility  Reactions  \
0         87.0         96.0          91.0         86.0     91.0       95.0   
1         77.0         94.0          89.0         91.0     87.0       96.0   
2         78.0         95.0          94.0         90.0     96.0       94.0   
3         51.0         42.0          57.0         58.0     60.0       90.0   
4         91.0         91.0          78.0         76.0     79.0       91.0   

   Balance  ShotPower  Jumping  Stamina  Strength  LongShots  Aggression  \
0     95.0       85.0     68.0     72.0      59.0       94.0        48.0   
1     70.0       95.0     95.0     88.0      79.0       93.0        63.0   
2     84.0       80.0     61.0     81.0      49.0       82.0        56.0   
3     43.0       31.0     67.0     43.0      64.0       12.0        38.0   
4     77.0       91.0     63.0     90.0      75.0       91.0        76.0   

   Interceptions  Positioning  Vision  Penalties  Composure  Marking  \
0           22.0         94.0    94.0       75.0       96.0     33.0   
1           29.0         95.0    82.0       85.0       95.0     28.0   
2           36.0         89.0    87.0       81.0       94.0     27.0   
3           30.0         12.0

Las columnas de **Value**, **Wage** y **Release Clause** ya se encuentran convertidas a las unidades correspondientes. Ahora hay que comprobar que el resto de los atributos no son cadenas de texto. Para ello usamos el comando `dtypes` de `panda`.

In [20]:
pd.set_option('display.max_rows', None)
df.dtypes

Age                           int64
Nationality                  object
Overall                       int64
Potential                     int64
Club                         object
Value                       float64
Wage                        float64
Preferred Foot               object
International Reputation    float64
Weak Foot                   float64
Skill Moves                 float64
Body Type                    object
Position                     object
Height                       object
Weight                       object
LS                           object
ST                           object
RS                           object
LW                           object
LF                           object
CF                           object
RF                           object
RW                           object
LAM                          object
CAM                          object
RAM                          object
LM                           object
LCM                         

La altura, el peso y diversas estadísticas no se encuentran en formato numérico por lo que hay que definir una serie de funciones para convertir estas cadenas de texto en valores con los que se pueda trabajar.

In [21]:
def height_to_float(x):
    
    ret_val = 0.0
    if chr(39) in x:
        ret_val = float(x.replace(chr(39), '.'))
    ret_val = float(ret_val)
    return ret_val

In [22]:
def weight_to_int(x):
    
    ret_val = 0
    if "lbs" in x:
        ret_val = int(x.replace("lbs", ''))
    ret_val = int(ret_val)
    return ret_val

In [23]:
def stats_to_int(x):
    if type(x) == float or type(x) == int:
        return x
    ret_val = 0
    if '+' in x:
        first = x.partition("+")[0] #obtener numero antes del signo de suma
        second = x.partition("+")[2] #obtener numero despues del signo de suma
        ret_val = int(first) + int(second)
    ret_val = int(ret_val)
    return ret_val

Ahora aplicamos las funciones sobre las columnas correspondientes. En el caso de las estadisticas se utiliza un bucle para no tener que repetir el mismo comando muchas veces.

In [24]:
pd.set_option('display.max_rows', 10)
df["Height"] = df["Height"].apply(height_to_float)
df["Weight"] = df["Weight"].apply(weight_to_int)
df.head()

Age Nationality  Overall  Potential                 Club        Value  \
0   31   Argentina       94         94         FC Barcelona  110500000.0   
1   33    Portugal       94         94             Juventus   77000000.0   
2   26      Brazil       92         93  Paris Saint-Germain  118500000.0   
3   27       Spain       91         93    Manchester United   72000000.0   
4   27     Belgium       91         92      Manchester City  102000000.0   

       Wage Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
0  565000.0           Left                       5.0        4.0          4.0   
1  405000.0          Right                       5.0        4.0          5.0   
2  290000.0          Right                       5.0        5.0          5.0   
3  260000.0          Right                       4.0        3.0          1.0   
4  355000.0          Right                       4.0        5.0          4.0   

    Body Type Position  Height  Weight    LS    ST    RS    LW    LF    CF  \
0       Messi       RF    5.70     159  88+2  88+2  88+2  92+2  93+2  93+2   
1  C. Ronaldo       ST    6.20     183  91+3  91+3  91+3  89+3  90+3  90+3   
2      Neymar       LW    5.90     150  84+3  84+3  84+3  89+3  89+3  89+3   
3        Lean       GK    6.40     168     0     0     0     0     0     0   
4      Normal      RCM    5.11     154  82+3  82+3  82+3  87+3  87+3  87+3   

     RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM   LWB   LDM  \
0  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2   
1  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3   
2  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3   
3     0     0     0     0     0     0     0     0     0     0     0     0   
4  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3   

    CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  Finishing  \
0  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0       95.0   
1  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0       94.0   
2  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3      79.0       87.0   
3     0     0     0     0     0     0     0     0      17.0       13.0   
4  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3      93.0       82.0   

   HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  \
0             70.0          90.0     86.0       97.0   93.0        94.0   
1             89.0          81.0     87.0       88.0   81.0        76.0   
2             62.0          84.0     84.0       96.0   88.0        87.0   
3             21.0          50.0     13.0       18.0   21.0        19.0   
4             55.0          92.0     82.0       86.0   85.0        83.0   

   LongPassing  BallControl  Acceleration  SprintSpeed  Agility  Reactions  \
0         87.0         96.0          91.0         86.0     91.0       95.0   
1         77.0         94.0          89.0         91.0     87.0       96.0   
2         78.0         95.0          94.0         90.0     96.0       94.0   
3         51.0         42.0          57.0         58.0     60.0       90.0   
4         91.0         91.0          78.0         76.0     79.0       91.0   

   Balance  ShotPower  Jumping  Stamina  Strength  LongShots  Aggression  \
0     95.0       85.0     68.0     72.0      59.0       94.0        48.0   
1     70.0       95.0     95.0     88.0      79.0       93.0        63.0   
2     84.0       80.0     61.0     81.0      49.0       82.0        56.0   
3     43.0       31.0     67.0     43.0      64.0       12.0        38.0   
4     77.0       91.0     63.0     90.0      75.0       91.0        76.0   

   Interceptions  Positioning  Vision  Penalties  Composure  Marking  \
0           22.0         94.0    94.0       75.0       96.0     33.0   
1           29.0         95.0    82.0       85.0       95.0     28.0   
2           36.0         89.0    87.0       81.0       94.0     27.0   
3           30.0       

In [25]:
stats = ['LS','ST','RS','LW','LF','CF','RF','RW','LAM','CAM','RAM','LM','LCM','CM','RCM','RM','LWB','LDM','CDM','RDM','RWB','LB','LCB','CB','RCB','RB']
for i in stats:
    df[i] = df[i].apply(stats_to_int)
df.head()

Age Nationality  Overall  Potential                 Club        Value  \
0   31   Argentina       94         94         FC Barcelona  110500000.0   
1   33    Portugal       94         94             Juventus   77000000.0   
2   26      Brazil       92         93  Paris Saint-Germain  118500000.0   
3   27       Spain       91         93    Manchester United   72000000.0   
4   27     Belgium       91         92      Manchester City  102000000.0   

       Wage Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
0  565000.0           Left                       5.0        4.0          4.0   
1  405000.0          Right                       5.0        4.0          5.0   
2  290000.0          Right                       5.0        5.0          5.0   
3  260000.0          Right                       4.0        3.0          1.0   
4  355000.0          Right                       4.0        5.0          4.0   

    Body Type Position  Height  Weight  LS  ST  RS  LW  LF  CF  RF  RW  LAM  \
0       Messi       RF    5.70     159  90  90  90  94  95  95  95  94   95   
1  C. Ronaldo       ST    6.20     183  94  94  94  92  93  93  93  92   91   
2      Neymar       LW    5.90     150  87  87  87  92  92  92  92  92   92   
3        Lean       GK    6.40     168   0   0   0   0   0   0   0   0    0   
4      Normal      RCM    5.11     154  85  85  85  90  90  90  90  90   91   

   CAM  RAM  LM  LCM  CM  RCM  RM  LWB  LDM  CDM  RDM  RWB  LB  LCB  CB  RCB  \
0   95   95  93   86  86   86  93   66   63   63   63   66  61   49  49   49   
1   91   91  91   84  84   84  91   68   64   64   64   68  64   56  56   56   
2   92   92  91   84  84   84  91   68   63   63   63   68  63   50  50   50   
3    0    0   0    0   0    0   0    0    0    0    0    0   0    0   0    0   
4   91   91  91   90  90   90  91   80   80   80   80   80  76   69  69   69   

   RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0  61      84.0       95.0             70.0          90.0     86.0       97.0   
1  64      84.0       94.0             89.0          81.0     87.0       88.0   
2  63      79.0       87.0             62.0          84.0     84.0       96.0   
3   0      17.0       13.0             21.0          50.0     13.0       18.0   
4  76      93.0       82.0             55.0          92.0     82.0       86.0   

   Curve  FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  \
0   93.0        94.0         87.0         96.0          91.0         86.0   
1   81.0        76.0         77.0         94.0          89.0         91.0   
2   88.0        87.0         78.0         95.0          94.0         90.0   
3   21.0        19.0         51.0         42.0          57.0         58.0   
4   85.0        83.0         91.0         91.0          78.0         76.0   

   Agility  Reactions  Balance  ShotPower  Jumping  Stamina  Strength  \
0     91.0       95.0     95.0       85.0     68.0     72.0      59.0   
1     87.0       96.0     70.0       95.0     95.0     88.0      79.0   
2     96.0       94.0     84.0       80.0     61.0     81.0      49.0   
3     60.0       90.0     43.0       31.0     67.0     43.0      64.0   
4     79.0       91.0     77.0       91.0     63.0     90.0      75.0   

   LongShots  Aggression  Interceptions  Positioning  Vision  Penalties  \
0       94.0        48.0           22.0         94.0    94.0       75.0   
1       93.0        63.0           29.0         95.0    82.0       85.0   
2       82.0        56.0           36.0         89.0    87.0       81.0   
3       12.0        38.0           30.0         12.0    68.0       40.0   
4       91.0        76.0           61.0         87.0    94.0       79.0   

   Composure  Marking  StandingTackle  SlidingTackle  GKDiving  GKHandling  \
0       96.0     33.0            28.0           26.0       6.0        11.0   
1       95.0     28.0            31.0           23.0       7.0        11.0   
2       94.0     27.0            24

## Gestión datos no numéricos

Por otra parte, nos quedan una serie de columnas como el **Club**, que no podemos manipular ya que se trata de una cadena de texto que no se puede convertir en un valor numérico. Para solucionar este problema aplicamos una función que lo que hace es crear una columna por cada Club que haya y así el valor de esa columna será 0 o 1 dependiendo de si el jugador pertenece al Club o no.

De la misma forma se procede con las demás columnas similares al **Club**. Las columnas afectadas son:
- Club
- Nationality
- Body Type
- Position
- Preferred Foot

In [26]:
pd.set_option('display.max_columns', 100)
clb = df.pop("Club")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(clb, prefix='clb').reset_index(drop=True)], axis=1, sort=False)
df.head()

Age Nationality  Overall  Potential        Value      Wage Preferred Foot  \
0   31   Argentina       94         94  110500000.0  565000.0           Left   
1   33    Portugal       94         94   77000000.0  405000.0          Right   
2   26      Brazil       92         93  118500000.0  290000.0          Right   
3   27       Spain       91         93   72000000.0  260000.0          Right   
4   27     Belgium       91         92  102000000.0  355000.0          Right   

   International Reputation  Weak Foot  Skill Moves   Body Type Position  \
0                       5.0        4.0          4.0       Messi       RF   
1                       5.0        4.0          5.0  C. Ronaldo       ST   
2                       5.0        5.0          5.0      Neymar       LW   
3                       4.0        3.0          1.0        Lean       GK   
4                       4.0        5.0          4.0      Normal      RCM   

   Height  Weight  LS  ST  RS  LW  LF  CF  RF  RW  LAM  CAM  RAM  LM  LCM  CM  \
0    5.70     159  90  90  90  94  95  95  95  94   95   95   95  93   86  86   
1    6.20     183  94  94  94  92  93  93  93  92   91   91   91  91   84  84   
2    5.90     150  87  87  87  92  92  92  92  92   92   92   92  91   84  84   
3    6.40     168   0   0   0   0   0   0   0   0    0    0    0   0    0   0   
4    5.11     154  85  85  85  90  90  90  90  90   91   91   91  91   90  90   

   RCM  RM  LWB  LDM  CDM  RDM  RWB  LB  LCB  CB  RCB  RB  Crossing  \
0   86  93   66   63   63   63   66  61   49  49   49  61      84.0   
1   84  91   68   64   64   64   68  64   56  56   56  64      84.0   
2   84  91   68   63   63   63   68  63   50  50   50  63      79.0   
3    0   0    0    0    0    0    0   0    0   0    0   0      17.0   
4   90  91   80   80   80   80   80  76   69  69   69  76      93.0   

   Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  \
0       95.0             70.0          90.0     86.0       97.0   93.0   
1       94.0             89.0          81.0     87.0       88.0   81.0   
2       87.0             62.0          84.0     84.0       96.0   88.0   
3       13.0             21.0          50.0     13.0       18.0   21.0   
4       82.0             55.0          92.0     82.0       86.0   85.0   

   FKAccuracy  LongPassing  BallControl  ...  clb_Unión de Santa Fe  \
0        94.0         87.0         96.0  ...                      0   
1        76.0         77.0         94.0  ...                      0   
2        87.0         78.0         95.0  ...                      0   
3        19.0         51.0         42.0  ...                      0   
4        83.0         91.0         91.0  ...                      0   

   clb_Urawa Red Diamonds  clb_V-Varen Nagasaki  clb_VVV-Venlo  \
0                       0                     0              0   
1                       0                     0              0   
2                       0                     0              0   
3                       0                     0              0   
4                       0                     0              0   

   clb_Valencia CF  clb_Valenciennes FC  clb_Vancouver Whitecaps FC  \
0                0                    0                           0   
1                0                    0                           0   
2                0                    0                           0   
3                0                    0                           0   
4                0                    0                           0   

   clb_Vegalta Sendai  clb_Vejle Boldklub  clb_Vendsyssel FF  clb_Venezia FC  \
0                   0                   0                  0               0   
1                   0                   0                  0               0   
2                   0                   0                  0               0   
3                   0                   0                  0               0   
4                   0                   0              

In [27]:
nat = df.pop("Nationality")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(nat, prefix='nat').reset_index(drop=True)], axis=1, sort=False)

In [28]:
bt = df.pop("Body Type")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(bt, prefix='bd').reset_index(drop=True)], axis=1, sort=False)

In [29]:
pos = df.pop("Position")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(pos, prefix='pos').reset_index(drop=True)], axis=1, sort=False)

In [30]:
pf = df.pop("Preferred Foot")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(pf, prefix='pref_foot').reset_index(drop=True)], axis=1, sort=False)
df.head()

Age  Overall  Potential        Value      Wage  International Reputation  \
0   31       94         94  110500000.0  565000.0                       5.0   
1   33       94         94   77000000.0  405000.0                       5.0   
2   26       92         93  118500000.0  290000.0                       5.0   
3   27       91         93   72000000.0  260000.0                       4.0   
4   27       91         92  102000000.0  355000.0                       4.0   

   Weak Foot  Skill Moves  Height  Weight  LS  ST  RS  LW  LF  CF  RF  RW  \
0        4.0          4.0    5.70     159  90  90  90  94  95  95  95  94   
1        4.0          5.0    6.20     183  94  94  94  92  93  93  93  92   
2        5.0          5.0    5.90     150  87  87  87  92  92  92  92  92   
3        3.0          1.0    6.40     168   0   0   0   0   0   0   0   0   
4        5.0          4.0    5.11     154  85  85  85  90  90  90  90  90   

   LAM  CAM  RAM  LM  LCM  CM  RCM  RM  LWB  LDM  CDM  RDM  RWB  LB  LCB  CB  \
0   95   95   95  93   86  86   86  93   66   63   63   63   66  61   49  49   
1   91   91   91  91   84  84   84  91   68   64   64   64   68  64   56  56   
2   92   92   92  91   84  84   84  91   68   63   63   63   68  63   50  50   
3    0    0    0   0    0   0    0   0    0    0    0    0    0   0    0   0   
4   91   91   91  91   90  90   90  91   80   80   80   80   80  76   69  69   

   RCB  RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0   49  61      84.0       95.0             70.0          90.0     86.0   
1   56  64      84.0       94.0             89.0          81.0     87.0   
2   50  63      79.0       87.0             62.0          84.0     84.0   
3    0   0      17.0       13.0             21.0          50.0     13.0   
4   69  76      93.0       82.0             55.0          92.0     82.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       18.0   21.0        19.0         51.0         42.0          57.0   
4       86.0   85.0        83.0         91.0         91.0          78.0   

   SprintSpeed  Agility  Reactions  ...  nat_Turkey  nat_Uganda  nat_Ukraine  \
0         86.0     91.0       95.0  ...           0           0            0   
1         91.0     87.0       96.0  ...           0           0            0   
2         90.0     96.0       94.0  ...           0           0            0   
3         58.0     60.0       90.0  ...           0           0            0   
4         76.0     79.0       91.0  ...           0           0            0   

   nat_United Arab Emirates  nat_United States  nat_Uruguay  nat_Uzbekistan  \
0                         0                  0            0               0   
1                         0                  0            0               0   
2                         0                  0            0               0   
3                         0                  0            0               0   
4                         0                  0            0               0   

   nat_Venezuela  nat_Wales  nat_Zambia  nat_Zimbabwe  bd_Akinfenwa  \
0              0          0           0             0             0   
1              0          0           0             0             0   
2              0          0           0             0             0   
3              0          0           0             0             0   
4              0          0           0             0             0   

   bd_C. Ronaldo  bd_Courtois  bd_Lean  bd_Messi  bd_Neymar  bd_Normal  \
0              0            0        0         1          0          0   
1              1            0        0         0          0          0   
2              0            0        0         0          1          0   
3  

## Predicción de valores

Una vez tenemos el conjunto de datos solucionado, sin *NaN* y con valores trabajables podemos realizar la predicción. Debemos predecir el valor de la columna **Value** a partir de los demás atributos.

Empezamos extrayendo la columna **Value** para poder aplicar la regresión.

In [31]:
val = df.pop("Value")
df.head()

Age  Overall  Potential      Wage  International Reputation  Weak Foot  \
0   31       94         94  565000.0                       5.0        4.0   
1   33       94         94  405000.0                       5.0        4.0   
2   26       92         93  290000.0                       5.0        5.0   
3   27       91         93  260000.0                       4.0        3.0   
4   27       91         92  355000.0                       4.0        5.0   

   Skill Moves  Height  Weight  LS  ST  RS  LW  LF  CF  RF  RW  LAM  CAM  RAM  \
0          4.0    5.70     159  90  90  90  94  95  95  95  94   95   95   95   
1          5.0    6.20     183  94  94  94  92  93  93  93  92   91   91   91   
2          5.0    5.90     150  87  87  87  92  92  92  92  92   92   92   92   
3          1.0    6.40     168   0   0   0   0   0   0   0   0    0    0    0   
4          4.0    5.11     154  85  85  85  90  90  90  90  90   91   91   91   

   LM  LCM  CM  RCM  RM  LWB  LDM  CDM  RDM  RWB  LB  LCB  CB  RCB  RB  \
0  93   86  86   86  93   66   63   63   63   66  61   49  49   49  61   
1  91   84  84   84  91   68   64   64   64   68  64   56  56   56  64   
2  91   84  84   84  91   68   63   63   63   68  63   50  50   50  63   
3   0    0   0    0   0    0    0    0    0    0   0    0   0    0   0   
4  91   90  90   90  91   80   80   80   80   80  76   69  69   69  76   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     86.0       97.0   
1      84.0       94.0             89.0          81.0     87.0       88.0   
2      79.0       87.0             62.0          84.0     84.0       96.0   
3      17.0       13.0             21.0          50.0     13.0       18.0   
4      93.0       82.0             55.0          92.0     82.0       86.0   

   Curve  FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  \
0   93.0        94.0         87.0         96.0          91.0         86.0   
1   81.0        76.0         77.0         94.0          89.0         91.0   
2   88.0        87.0         78.0         95.0          94.0         90.0   
3   21.0        19.0         51.0         42.0          57.0         58.0   
4   85.0        83.0         91.0         91.0          78.0         76.0   

   Agility  Reactions  Balance  ...  nat_Turkey  nat_Uganda  nat_Ukraine  \
0     91.0       95.0     95.0  ...           0           0            0   
1     87.0       96.0     70.0  ...           0           0            0   
2     96.0       94.0     84.0  ...           0           0            0   
3     60.0       90.0     43.0  ...           0           0            0   
4     79.0       91.0     77.0  ...           0           0            0   

   nat_United Arab Emirates  nat_United States  nat_Uruguay  nat_Uzbekistan  \
0                         0                  0            0               0   
1                         0                  0            0               0   
2                         0                  0            0               0   
3                         0                  0            0               0   
4                         0                  0            0               0   

   nat_Venezuela  nat_Wales  nat_Zambia  nat_Zimbabwe  bd_Akinfenwa  \
0              0          0           0             0             0   
1              0          0           0             0             0   
2              0          0           0             0             0   
3              0          0           0             0             0   
4              0          0           0             0             0   

   bd_C. Ronaldo  bd_Courtois  bd_Lean  bd_Messi  bd_Neymar  bd_Normal  \
0              0            0        0         1          0          0   
1              1            0        0         0          0          0   
2              0            0        0         0          1          0   
3              0           

Para empezar a predecir hay que separar los datos para poder conocer su rendimiento una vez finalizado el proceso. Aplicamos la función `train_test_split` de la libreria `sklearn`.

In [32]:
X_train, X_test, y_train, y_test = train_test_split(df, val, test_size=0.33, random_state=42)

In [33]:
len(X_train)

12005

Ahora hay que entrenar el modelo usando la regresión lineal vista en clase. Procedemos con el siguiente comando:

In [34]:
reg = linear_model.LinearRegression().fit(X_train, y_train)

In [35]:
preds = reg.predict(X_test)

In [36]:
preds[1]

1335137.0282127308

In [37]:
y_test[0]

110500000.0

Para acabar obtenemos la métrica $R^2$ para la regresión. Hemos hecho uso de la implementación de la libreria `sklearn`.

In [38]:
final_score = r2_score(preds, y_test)

In [39]:
print('El porcentaje de exactitud a la hora de predecir el valor de un jugador es de',format(final_score*100,'.2f'),'%')

El porcentaje de exactitud a la hora de predecir el valor de un jugador es de 96.56 %


## Conclusión

Podemos determinar que la predicción aplicada es bastante acertada ya que el porcentaje de exactitud es superior al 96%.